src/lib.rs

```diff
@@ -40,19 +42,358 @@ mod repl_helper;
+        ///
+        /// # Expected Response Format
+        /// ```json
+        /// {
+        ///   "status": "ok",  // Can also be "error" or "panic"
+        ///   "result": [ // Only present if status is "ok"
+        ///     { ... }, // Each is a Map representing a query result.
+        ///   ],
+        ///   "error": "..." // Only present if status is "error"
+        /// }
+        #[no_mangle]
+        pub unsafe extern "C" fn scryer_run_query(machine: &mut Machine, input: *const c_char) -> *mut c_char {
+            let c_string;
+            let r_str;
+            unsafe {
+                c_string = CStr::from_ptr(input);
+                r_str = c_string.to_str().expect("Not a valid UTF-8 string");
+            };
+            let result = std::panic::catch_unwind(|| r_str);
+
+            let output_string = match result {
+                Ok(r_str) => {
+                    let query_resolution = machine.run_query(r_str.to_owned());
+                    let value: serde_json::Value = serde_json::from_str(&format!("{}", query_resolution.expect("Something went wrong marshaling JSON"))).unwrap();
+                    serde_json::to_string(&serde_json::json!({"status": "ok", "result": value})).unwrap()
+                }
+                Err(e_str) => serde_json::to_string(&serde_json::json!({"status": "error", "error": format!("{:?}", &e_str)})).unwrap()
+            };
 
-#[cfg(target_arch = "wasm32")]
-use wasm_bindgen::prelude::*;
+            let c_string = CString::new(output_string).unwrap();
+            c_string.into_raw()
+        }
 
-#[cfg(target_arch = "wasm32")]
-#[wasm_bindgen]
-pub fn eval_code(s: &str) -> String {
-    use machine::mock_wam::*;
 
-    console_error_panic_hook::set_once();
+        #[no_mangle]
+        /// This function is responsible for freeing up the memory that was previously allocated for a C
+        /// string that was returned by Scryer Prolog binding functions. It is of critical importance to
+        /// prevent memory leaks that this function has to be invoked after every single invocation of
+        /// the following functions: `scryer_consult_module_string()`, `scryer_load_module_string()`,
+        /// `scryer_run_query()`, `scryer_run_query_generator()`, `scryer_start_new_query_generator()`,
+        /// and `scryer_cleanup_query_generator()`.
+        ///
+        /// # Arguments
+        ///
+        /// * `ptr: *mut c_char` - A mutable pointer to c_char which indicates the memory address that should be freed.
+        ///
+        /// # Safety
+        ///
+        /// This function runs a unsafe block of code. The "unsafe" keyword is used because this function
+        /// makes use of a raw pointer and the Rust memory safety guarantees cannot be upheld. Please
+        /// ensure to follow the usage instructions to prevent any unintended side effects.
+        pub unsafe extern "C" fn scryer_free_c_string(ptr: *mut c_char) {
+            if ptr.is_null() {
```

Comment:
> It is of critical importance to prevent memory leaks that this function has to be invoked after every single invocation of the following functions

This makes it sound like it needs to happen directly after the functions are called, but the program may hold onto them and deallocate them at a later point. 
Also this should probably be called out on the listed functions rather than here as the user is likely to read the documentation of a function they are using rather then that of random other function.

The safety section in the doc comment if for the caller to tell them the safety requirements they need to uphold. It is supposed to contain the usage instructions relevant for safety rather than point to them.